lib/perl/Genome/Model/Tools/Bsmap/MethCalcConversionRate.pm

```diff
@@ -0,0 +1,263 @@
+
+EOS
+}
+
+sub help_brief {
+    "calculate methylation conversion using Mitochondrial DNA or spiked lambda"
+}
+
+sub help_detail {
+    "calculate methylation conversion using Mitochondrial DNA or spiked lambda"
+}
+
+sub bs_rate {
+    my $count = 0;
+    my $totalreads = 0;
+    my $methreads = 0;
+
+    open (my $fh, '<', $_[0]) or die("Could not open snvs file.");
+    while (my $line = <$fh>) {
+        if ( $count == 0 ) {
+            $count = 1;
+            next;
+        }
+        chomp($line);
+        my @F = split("\t",$line);
+        if(($F[2] eq "-" && $F[3] =~ /.CG../ ) || ($F[2] eq "+" && $F[3] =~ /..CG./)){
+            $totalreads = $totalreads + $F[5];
+            $methreads = $methreads + $F[6];
+        }
+    }
+    close($_[0]);
+
+    my $cfile;
+
+    if (defined($_[2])){
+        open($cfile, ">>$_[2]") || die "Can't open output file for writing.\n";
+        if($_[1] eq "MT"){
+            print $cfile "\nMethylation conversion based on mtDNA:\n";
+        }
+        if($_[1] eq "lambda"){
+            print $cfile "\nMethylation conversion based on lambda:\n";
+        }
+        print $cfile "Meth reads\t=\t", $methreads, "\n";
+        print $cfile "Total reads\t=\t", $totalreads, "\n";
+        if ($totalreads != 0) {
+            print $cfile "Bisulfite conversion (%)\t=\t", 100-($methreads/$totalreads*100), "\n\n";
+        }
+        close $cfile;
+    }
+    else {
+        print "Meth reads\t=\t", $methreads, "\n";
+        print "Total reads\t=\t", $totalreads, "\n";
+        if ($totalreads != 0) {
+            print "Bisulfite conversion (%)\t=\t", 100-($methreads/$totalreads*100), "\n\n";
+        }
+    }
+}
+
+sub execute {
+    my $self = shift;
+    my $snvs_file = $self->snvs_file;
+    my $model_id = $self->model_id;
+    my $output_file = $self->output_file;
+
+    if (!defined($snvs_file) && !defined($model_id) ){
+        die("Must provide snvs file OR model ID.\n");
+    }
+
+    if (defined($snvs_file) && defined($model_id) ){
+        die("Must provide snvs file OR model ID.\n");
+    }
+
+    my $cfile;
+
+    if (defined($snvs_file)){
+        # snvs
+        if (-s "$snvs_file"){
+            if (defined($output_file)){
+                open($cfile, ">$output_file") || die "Can't open output file for writing.\n";
+                close($cfile);
+                bs_rate($snvs_file, "MT", $output_file);
+            }
+            else {
+                bs_rate($snvs_file, "MT");
+            }
+        }
+        else {
+            $self->error_message("can't find the snvs file");
+        }
+    }
+
+    if (defined($model_id)){
+        # get model IDs
+        my $model = Genome::Model->get($model_id);
+        # get the bam paths of the last succeeded build
+        my $dir = $model->last_succeeded_build->data_directory;
+        # flagstat 
+        my @flagstat = glob("$dir/alignments/*flagstat");
```

Comment:
It is better to ask the build for its flagstat file that to reach in to the filesystem.  For example, `Genome::Model::Build::ReferenceAlignment::whole_rmdup_bam_flagstat_file()`.